### ISSUE TITLE
自己的数据集运行DSH很慢

### ISSUE THREAD

**Issue by 974947823** [swuxyj/DeepHash-pytorch#29]
`class DSHLoss(torch.nn.Module):
    def __init__(self, config, bit):
        super(DSHLoss, self).__init__()
        self.m = 2 * bit
        self.U = torch.zeros(config["num_train"], bit).float().to(config["device"])
        self.Y = torch.zeros(config["num_train"], config["n_class"]).float().to(config["device"])

    def forward(self, u, y, ind, config):
        self.U[ind, :] = u.data
        self.Y[ind, :] = y.float()

        dist = (u.unsqueeze(1).half() - self.U.unsqueeze(0).half()).pow(2).sum(dim=2)
        y = (y @ self.Y.t() == 0).float()

        loss = (1 - y) / 2 * dist + y / 2 * (self.m - dist).clamp(min=0)
        loss1 = loss.mean()
        loss2 = config["alpha"] * (1 - u.abs()).abs().mean()

        return loss1 + loss2
`
`
def hashing_loss(b, cls, m, alpha):
    """
    compute hashing loss
    automatically consider all n^2 pairs
    """
    y = (cls.unsqueeze(0) != cls.unsqueeze(1)).float().view(-1)
    dist = ((b.unsqueeze(0) - b.unsqueeze(1)) ** 2).sum(dim=2).view(-1)
    loss = (1 - y) / 2 * dist + y / 2 * (m - dist).clamp(min=0)

    loss = loss.mean() + alpha * (b.abs() - 1).abs().sum(dim=1).mean() * 2

    return loss`

第一个是您代码中Loss，第二个是
[DSH-pytorch](https://github.com/weixu000/DSH-pytorch)的实现，第二个中并没有生成shape为[num_train,hashbit]的向量，这会不会让训练速度变慢呢

**Comment by swuxyj** [comment]
训练的耗时主要在于神经网络那部分的计算，这里的矩阵运算非常快，你可以对比一下两种写法的耗时，应该是差不多的哈

**Comment by 974947823** [comment]
> 
self.U = torch.zeros(config["num_train"], bit).float().to(config["device"])
self.Y = torch.zeros(config["num_train"], config["n_class"]).float().to(config["device"])
这两行代码主要的目的是什么呢，我这里有60多万张图片，建立Index就已经显存炸裂了。新手，求解答

**Comment by swuxyj** [comment]
我看到很多代码都是用的这种写法，应该就是存储训练集的特征，用来计算损失  
如果显存不够，那就删掉直接算batch_size × batch_size 的数据吧，类似这样写就行：
```
dist = (u.unsqueeze(1) - u.unsqueeze(0)).pow(2).sum(dim=2)
y = (y @ y.t() == 0).float()
```
